kernel/libs/xarray/src/lib.rs

```diff
@@ -0,0 +1,249 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! This crate introduces the RCU-based xarray.
+
+#![no_std]
+#![deny(unsafe_code)]
+
+extern crate alloc;
+
+use core::marker::PhantomData;
+
+pub use cursor::{Cursor, CursorMut};
+use entry::XEntry;
+use mark::NoneMark;
+pub use mark::XMark;
+use ostd::{
+    sync::{
+        non_null::NonNullPtr, LocalIrqDisabled, PreemptDisabled, RcuOption, SpinGuardian, SpinLock,
+        SpinLockGuard,
+    },
+    task::atomic_mode::AsAtomicModeGuard,
+};
+pub use range::Range;
+
+mod cursor;
+mod entry;
+mod mark;
+mod node;
+mod range;
+
+#[cfg(ktest)]
+mod test;
+
+const BITS_PER_LAYER: usize = 6;
+const SLOT_SIZE: usize = 1 << BITS_PER_LAYER;
+const SLOT_MASK: usize = SLOT_SIZE - 1;
+
+/// `XArray` is an abstract data type functioning like an expansive array of items
```

Comment:
I think a large portion of this type-level Rust doc should be moved to the module-level doc. People need to get an overview of what this crate has to offer.